### PR TITLE
feat: add confirmLastLook request to MMs

### PR DESCRIPTION
**Pull request by phil-ociraptor**
# Description

# Testing Instructions

# Checklist

-   [ ] Update [documentation](https://github.com/0xProject/website/blob/development/mdx/api/index.mdx) as needed. **Website Documentation PR:**
-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Test changes on the staging environment with [Matcha API Staging](https://api-staging.matcha.xyz).

    -   [ ] SRA/Limit orders
    -   [ ] Swap endpoints
    -   [ ] Meta transaction endpoints
    -   [ ] Depth charts

    For more information see `0x API Matcha smoke test runbook` in Quip.

### CODE REVIEW

File: src/utils/quote_server_client.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    name: 'market_maker_last_look_latency',
+    help: 'Latency for Last Look request to Market Makers',
+    labelNames: ['makerUri'],
+});
+
+export class QuoteServerClient {
+    private readonly _schemaValidator: SchemaValidator;
+
+    private static _areFeesEqual(fee1: Fee, fee2: Fee): boolean {
+        return fee1.amount.eq(fee2.amount) && fee1.token === fee2.token && fee1.type === fee2.type;
```
Comment (PirosB3) [inline src/utils/quote_server_client.ts:21]: Why not use [LoDash deep equal?](https://lodash.com/docs/4.17.15#isEqual)

File: src/utils/quote_server_client.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    public async confirmLastLookAsync(makerUri: string, payload: SubmitRequest): Promise<boolean> {
+        const timerStopFn = MARKET_MAKER_LAST_LOOK_LATENCY.labels(makerUri).startTimer();
+        try {
+            const response = await this._axiosInstance.post(`${makerUri}/submit`, payload, {
+                timeout: RFQT_REQUEST_MAX_RESPONSE_MS,
+            });
+            this._schemaValidator.validate(response.data, schemas.submitReceiptSchema);
+            const responseFee: Fee = {
```
Comment (PirosB3) [inline src/utils/quote_server_client.ts:38]: Why not simply compare the string values? without having to convert to `BigNumber`?

File: src/utils/quote_server_client.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        this._schemaValidator.addSchema(schemas.feeSchema);
+        this._schemaValidator.addSchema(schemas.submitRequestSchema);
+        this._schemaValidator.addSchema(schemas.submitReceiptSchema);
+    }
+
+    public async confirmLastLookAsync(makerUri: string, payload: SubmitRequest): Promise<boolean> {
+        const timerStopFn = MARKET_MAKER_LAST_LOOK_LATENCY.labels(makerUri).startTimer();
+        try {
+            const response = await this._axiosInstance.post(`${makerUri}/submit`, payload, {
+                timeout: RFQT_REQUEST_MAX_RESPONSE_MS,
```
Comment (PirosB3) [inline src/utils/quote_server_client.ts:35]: I would say we can be generous on timeout here, right? we are not "in a rush" and we could potentially benefit in a few more seconds timeout just to give makers all the opportunity possible to respond. Wouldn't you say?

File: src/utils/quote_server_client.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                amount: new BigNumber(response.data.fee.amount),
+                token: response.data.fee.token,
+                type: response.data.fee.type,
+            };
+
+            if (!QuoteServerClient._areFeesEqual(responseFee, payload.fee)) {
+                throw new Error('Fee in response is not equal to fee in request');
+            }
+
+            return response.data.proceedWithFill;
```
Comment (PirosB3) [inline src/utils/quote_server_client.ts:48]: Just to be extra safe, can we do `return response.data.proceedWithFill === true`?

File: src/utils/quote_server_client.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+export class QuoteServerClient {
+    private readonly _schemaValidator: SchemaValidator;
+
+    private static _areFeesEqual(fee1: Fee, fee2: Fee): boolean {
+        return fee1.amount.eq(fee2.amount) && fee1.token === fee2.token && fee1.type === fee2.type;
+    }
+
+    constructor(private readonly _axiosInstance: AxiosInstance) {
+        this._schemaValidator = new SchemaValidator();
```
Comment (PirosB3) [inline src/utils/quote_server_client.ts:25]: Why not initialize `schemaValidator` at a module level? (like you did for `MARKET_MAKER_LAST_LOOK_LATENCY`)? just a suggestion

**Review by PirosB3 — APPROVED** [review]
This is awesome stuff!

**Comment by PirosB3** [comment]
Just added a few minor comments

File: src/utils/quote_server_client.ts (hunk 2 above)
Comment (phil-ociraptor) [reply src/utils/quote_server_client.ts:38]: so the `payload` is going to have a `fee` property of type `Fee`. I'm converting the response here to BigNumber such that it conforms to type `Fee` as well. This way comparison can occur on two objects of type `Fee`.

File: src/utils/quote_server_client.ts (hunk 1 above)
Comment (phil-ociraptor) [reply src/utils/quote_server_client.ts:21]: will give that a shot! Could remove some code if so

File: src/utils/quote_server_client.ts (hunk 3 above)
Comment (phil-ociraptor) [reply src/utils/quote_server_client.ts:35]: 👍  increased to 2 seconds

**Pull request merged**